jax/_src/lax/lax.py

```diff
@@ -113,6 +127,7 @@ def _broadcast_shapes_cached(*shapes: Tuple[int, ...]) -> Tuple[int, ...]:
   return _broadcast_shapes_uncached(*shapes)
 
 def _broadcast_shapes_uncached(*shapes):
+  _validate_shapes(shapes)
```

Comment:
I'm not sure but I think an alternative here would be to only call `_validate_shapes(shapes)` in `_broadcast_shapes_cached` above (rather than here in `_broadcast_shapes_uncached`), and then we could even delete the `if config.jax_dynamic_shapes` logic in `_validate_shapes`. The idea is that the `try`/`except` in `broadcast_shapes` will fail (in `_validate_shapes`) only if dynamic shapes are involved.

But also this approach LGTM. Just thought I'd mention that possible alternative.

Reply:
I just tried this locally, and unfortunately it restores the old behavior. But the dynamic shape test passes, just like you hinted. I think we still need to employ the validation machinery in the uncached code path to account for cache population / misses.